codex/merkletree/merkletree.nim

```diff
@@ -1,96 +1,136 @@
-import pkg/questionable
 import pkg/questionable/results
+import pkg/nimcrypto/sha2
+
+const digestSize = sha256.sizeDigest
 
 type
-  MerkleHash* = MultiHash
+  MerkleHash* = array[digestSize, byte]
   MerkleTree* = object
-    leavesCount: int
+    leavesCount: Natural
     nodes: seq[MerkleHash]
   MerkleProof* = object
-    index: int
+    index: Natural
     path: seq[MerkleHash]
+  MerkleTreeBuilder* = object
+    buffer: seq[MerkleHash]
 
-# Tree constructed from leaves H0..H2 is
-#     H5=H(H3 & H4)
-#    /              \
-#   H3=H(H0 & H1)   H4=H(H2 & H2)
-#  /      \        /
-# H0=H(A) H1=H(B) H2=H(C)
-# |       |       |
-# A       B       C
-#
-# Memory layout is [H0, H1, H2, H3, H4, H5]
-#
-# Proofs of inclusion are
-# - [H1, H4] for A
-# - [H0, H4] for B
-# - [H2, H3] for C
-
+###########################################################
+# Helper functions
+###########################################################
 
 func computeTreeHeight(leavesCount: int): int =
   if isPowerOfTwo(leavesCount): 
     fastLog2(leavesCount) + 1
   else:
     fastLog2(leavesCount) + 2
 
-func getLowHigh(leavesCount, level: int): (int, int) =
-  var width = leavesCount
-  var low = 0
-  for _ in 0..<level:
-    low += width
-    width = (width + 1) div 2
+func computeLevels(leavesCount: int): seq[tuple[offset: int, width: int]] =
+  let height = computeTreeHeight(leavesCount)
+  result = newSeq[tuple[offset: int, width: int]](height)
+
+  result[0].offset = 0
+  result[0].width = leavesCount
+  for i in 1..<height:
+    result[i].offset = result[i - 1].offset + result[i - 1].width
+    result[i].width = (result[i - 1].width + 1) div 2
+
+proc digestFn(data: openArray[byte], output: var MerkleHash): void =
+  var digest = sha256.digest(data)
+  copyMem(addr output, addr digest.data[0], digestSize)
+
+###########################################################
+# MerkleHash
+###########################################################
+
+var zeroHash: MerkleHash
+
+proc `$`*(self: MerkleHash): string = 
+  result = newStringOfCap(self.len)
+  for i in 0..<self.len:
+    result.add(toHex(self[i]))
+
+###########################################################
+# MerkleTreeBuilder
+###########################################################
+
+proc addDataBlock*(self: var MerkleTreeBuilder, dataBlock: openArray[byte]): void =
+  ## Hashes the data block and adds the result of hashing to a buffer
+  ## 
+  let oldLen = self.buffer.len
+  self.buffer.setLen(oldLen + 1)
+  digestFn(dataBlock, self.buffer[oldLen])
+
+proc build*(self: MerkleTreeBuilder): ?!MerkleTree =
+  ## Builds a tree from previously added data blocks
+  ## 
+  ## Tree built from data blocks A, B and C is
+  ##        H5=H(H3 & H4)
+  ##      /            \
+  ##    H3=H(H0 & H1)   H4=H(H2 & HZ)
+  ##   /    \          /
+  ## H0=H(A) H1=H(B) H2=H(C)
+  ## |       |       |
+  ## A       B       C
+  ##
+  ## where HZ=H(0x0b)
+  ##
+  ## Memory layout is [H0, H1, H2, H3, H4, H5]
```

Comment:
Adressing the [comment](https://github.com/codex-storage/nim-codex/pull/504#discussion_r1287730466) from previous PR from @elcritch 
 
I added some docs to individual methods instead of a module.